libraries/Microsoft.Bot.Builder.AI.QnA/Dialogs/QnAMakerDialog.cs

```diff
@@ -0,0 +1,391 @@
+                // -Wait for the reply
+                // -If no: Skip to next step
+
+                var answer = response.First();
+
+                if (answer.Context != null && answer.Context.Prompts.Count() > 0)
+                {
+                    var previousContextData = ObjectPath.GetPathValue(stepContext.ActiveDialog.State, QnAContextData, new Dictionary<string, int>());
+                    var previousQnAId = ObjectPath.GetPathValue<int>(stepContext.ActiveDialog.State, PreviousQnAId, 0);
+
+                    foreach (var prompt in answer.Context.Prompts)
+                    {
+                        previousContextData.Add(prompt.DisplayText, prompt.QnaId);
+                    }
+
+                    ObjectPath.SetPathValue(stepContext.ActiveDialog.State, QnAContextData, previousContextData);
+                    ObjectPath.SetPathValue(stepContext.ActiveDialog.State, PreviousQnAId, answer.Id);
+                    ObjectPath.SetPathValue(stepContext.ActiveDialog.State, Options, dialogOptions);
+
+                    // Get multi-turn prompts card activity.
+                    var message = QnACardBuilder.GetQnAPromptsCard(answer, dialogOptions.ResponseOptions.CardNoMatchText);
+                    await stepContext.Context.SendActivityAsync(message).ConfigureAwait(false);
+
+                    return new DialogTurnResult(DialogTurnStatus.Waiting);
+                }
+            }
+
+            return await stepContext.NextAsync(stepContext.Result, cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task<DialogTurnResult> DisplayQnAResultAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
+        {
+            var dialogOptions = ObjectPath.GetPathValue<QnAMakerDialogOptions>(stepContext.ActiveDialog.State, Options);
+            var reply = stepContext.Context.Activity.Text;
+
+            if (reply.Equals(dialogOptions.ResponseOptions.CardNoMatchText, StringComparison.OrdinalIgnoreCase))
+            {
+                var activity = dialogOptions.ResponseOptions.CardNoMatchResponse;
+                if (activity == null)
+                {
+                    await stepContext.Context.SendActivityAsync(DefaultCardNoMatchResponse, cancellationToken: cancellationToken).ConfigureAwait(false);
+                }
+                else
+                {
+                    await stepContext.Context.SendActivityAsync(activity, cancellationToken: cancellationToken).ConfigureAwait(false);
+                }
+
+                return await stepContext.EndDialogAsync().ConfigureAwait(false);
+            }
+
+            // If previous QnAId is present, replace the dialog
+            var previousQnAId = ObjectPath.GetPathValue<int>(stepContext.ActiveDialog.State, PreviousQnAId, 0);
+            if (previousQnAId > 0)
+            {
+                return await stepContext.ReplaceDialogAsync(this.Id, dialogOptions, cancellationToken).ConfigureAwait(false);
```

Comment:
Instead of dialogOptions, we should send stepContext.ActiveDialog.State[Options]

Reply:
PreviousContext data is not being sent when we send dialogOptions

Reply:
The dialogOptions as defined IS stepContext.ActivityDialog.State[Options] only using ObjectPath to get the value as a strongly typed structure.

---
In reply to: [352744709](https://github.com/microsoft/botbuilder-dotnet/pull/3048#discussion_r352744709) [](ancestors = 352744709)

Reply:
Sorry, my bad. Instead of sending only Options value, I was asking to send entire stepContextActivityDialog.State. 
Currently n replace dialog, we are passing qnamakerOptions and response options but we are not passing previousQnAId and previousContextData, which gets cleared when replace dialog is called. That is the issue why tests was failing and you had to pass context as null.

Reply:
thanks, that indeed fixed the old unit test, I have restored the unit test and it passes now